libqtile/widget/decorations.py

```diff
@@ -0,0 +1,227 @@
+# Copyright (c) 2021, elParaguayo. All rights reserved.
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+import math
+from typing import Any, List, Tuple
+
+from libqtile.log_utils import logger
+from libqtile.widget import base
+
+
+class _Decoration(base.PaddingMixin):
+    """
+        Base decoration class. Should not be called by
+        configs directly.
+    """
+
+    defaults = [
+        ("padding", 0, "Default padding")
+    ]  # type: List[Tuple[str, Any, str]]
+
+    def __init__(self, **config):
+        base.PaddingMixin.__init__(self, **config)
+        self.add_defaults(_Decoration.defaults)
+
+    @property
+    def height(self):
+        return self.parent.height
+
+    @property
+    def width(self):
+        return self.parent.width
+
+    @property
+    def drawer(self):
+        return self.parent.drawer
+
+    @property
+    def ctx(self):
+        return self.parent.drawer.ctx
+
+    def draw(self, parent):
+        self.parent = parent
```

Comment:
Is it necessary to set the parent each time? Could it change?

I wonder if the base class shouldn't have a `_configure` method, called by the `Bar` along with the widgets, to give the decoration the parent widget. Then the draw method here can be an abstractmethod, ensuring nobody instantiates this base class directly. And also means child classes don't need to call `_Decoration.draw`.

Reply:
So... the way it works currently is that the decoration is instantiated once in the config but can be attached to multiple widgets (with the benefit of preserving the configuration of the decoration). However, at that point, it's not aware of any of the widgets' properties as they haven't been configured. When the widgets draw themselves, the decoration needs to know how big the parent widget is and so needs a reference to that parent. This means the same `draw` method is called for different widgets, hence the need to set the `parent`.

I thought about having a `_configure` method but I think that means each widget needs a separate instance of the decoration.

Reply:
From a config perspective, these aren't mutually exclusive ideas; a `_Decoration.copy()` could be used internally to give each widget its own instance so that the decorations each have exactly one parent and configs can still pass one instance to widgets.

So I guess the decision then comes down to performance or code clarity... not really sure what's the best there so happy for you to choose whichever.